grep-0018-modernize-c++.md

```diff
@@ -0,0 +1,417 @@
+- 20-Jan-2020: Initial Draft
+
+## Abstract
+
+As of early 2020 there is lots of pre-C++11 code in GNU Radio, and
+some other pattern of non-idiomatic C++.
+
+This GREP lays out how to get from here to there with minimal
+breaking.
+
+## Copyright / License
+
+This GREP is licensed as CC-BY-ND. Copyright 2020 Google Inc.
+
+## Motivation
+
+Cleaning up whole patterns of the code 
+
+The legacy patterns this GREP lays out a plan for making the code base:
+
+* more readable: standard components, less code, and explicit const
+* safer: less manual memory management, prevent inherently broken copy
+* faster: fewer pointer indirections, more information to the compiler
+
+## Description
+
+### Open questions
+
+* What level of ABI guarantee does GNU Radio make between and within
+  minor releases?
+
+### More const
+
+Benefits:
+* more readable; even in a function reader can see that a name has a
+  final value
+* avoids accidents
+* possibly allows compiler optimizations
+
+Recommendation:
+* private members and other `_impl.xx` code should be free to change
+  at any time to add const
+* adding const to exposed interfaces can be done between minor
+  releases
+
+Risks: none.
+
+### More constexpr
+
+`constexpr` guarantees compile time evaluation.
+
+Benefits over const:
+* no risk of initialization order breakage
+* can be used in some places where const can't
+* enables compiler optimizations
+
+Recommendation: See "More const", above
+
+Risks: none.
+
+### Depointerization
+
+GNU Radio has a lot of member variables that are raw pointers. In many
+cases ([example][raw-pointer-to-obj]) they don't have to be pointers
+at all, which:
+* clarifies for the reader what the lifetime of the object is
+* removes pointer dereference at runtime
+* possibly enables compiler to make smarter optimizations
+* reduces need for manual memory management, and in many cases the
+  whole destructor
+
+Other raw pointers do need to remain dynamic since changing them
+requires instantiating a new object
+([example][raw-pointer-to-std]). They should be changed to `std` smart
+pointers, or even more preferably the objects should be movable and be
+changed that way. For objects where copies are cheap it's enough to be
+able to copy.
+
+Another class of raw pointers is arrays, which sometimes require
+volk-specific allocation. Now that `volk::vector` exists these can all
+be changed to `std::vector` or `volk::vector`.
+
+Recommendation:
+* Allow 3.9 release to break ABI by changing raw and Boost pointers to
+  `std` and non-pointers even in public interfaces.
+
+Risks: none.
+
+### Delete copy constructor / copy assignment
+
+Where classes are not copyable (e.g. contain a pointer), the copy
+constructor and copy assignment operator should be deleted.
+
+```c++
+  foo(const foo&) = delete;
+  foo& operator(const foo&) = delete;
+```
+
+This should be added even if the class has a `std::unique_ptr` member
+variable, since it may be changed to `std::shared_ptr` in the future.
```

Comment:
If a class contains a shared_ptr, why not copy it?

Reply:
Fair enough. Deleted.